ENT-11255: Scan attachments to determine if they are Kotlin 1.2 or later

shamsasari opened the pull request:
The node now sends a transaction to the verifier if any of its attachments were compiled with Kotlin 1.2 (the `net.corda.node.verification.external` system property has been removed). It uses `kotlinx-metadata` to read the Kotlin metadata in the attachment to determine this. For now this scanning is done each time the attachment is loaded from the database.

The existing external verification integration tests were converted into smoke tests so that 4.11 nodes could be involved. This required various improvements to `NodeProcess.Factory`. A new `JAVA_8_HOME` environment variable, pointing to JDK 8, is required to run these tests.

There is still some follow-up work that needs to be done:

* Sending transactions from a 4.11 node to a 4.12 node works, but not the other way round. A new `WireTransaction` component group needs to be [introduced](https://r3-cev.atlassian.net/browse/ENT-11355) for storing 4.12 attachments so that they can be safely ignored by 4.11 nodes, and the 4.12 node needs to be able to load both 4.11 and 4.12 versions of the same contracts CorDapp so that they can be both attached to the transaction.
* Even though attachments are cached when retrieved from the database, the Kotlin metadata version should be [stored](https://r3-cev.atlassian.net/browse/ENT-11356) in the attachments db table, rather than being scanned each time.


Finally, `VerificationService` was refactored into `NodeVerificationSupport` and can be passed into `SignedTransaction.verifyInternal`, instead of needing the much heavier `VerifyingServiceHub`. This makes it easier for internal tools to verify transactions and spawn the verifier if necessary.

rick-r3 commented on node/src/main/kotlin/net/corda/node/verification/ExternalVerifierHandleImpl.kt:
```diff
@@ -188,20 +193,20 @@ class ExternalVerifierHandle(private val serviceHub: ServiceHubInternal) : AutoC
             verifierProcess.onExit().whenComplete { _, _ ->
                 if (connection != null) {
-                    log.error("The external verifier has unexpectedly terminated with error code ${verifierProcess.exitValue()}. " +
+                    log.warn("The external verifier has unexpectedly terminated with error code ${verifierProcess.exitValue()}. " +
```
It might be worth filing away a JIRA ticket to look at retry logic around a dead external verifier process.  e.g. if you just kill the external verifier it _should_ spring back into life, and we shouldn't fail a verify unless it happens repeatedly.  This is a new possibility from the introduction of the external verifier.  Very much an edge case but JVMs can crash (or OOM).

shamsasari replied:
https://r3-cev.atlassian.net/browse/ENT-11360

shamsasari committed:
ENT-11255: Scan attachments to determine if they are Kotlin 1.2 or later

The node now sends a transaction to the verifier if any of its attachments were compiled with Kotlin 1.2 (the net.corda.node.verification.external system property has been removed). It uses kotlinx-metadata to read the Kotlin metadata in the attachment to determine this. For now this scanning is done each time the attachment is loaded from the database.

The existing external verification integration tests were converted into smoke tests so that 4.11 nodes could be involved. This required various improvements to NodeProcess.Factory. A new JAVA_8_HOME environment variable, pointing to JDK 8, is required to run these tests.

There is still some follow-up work that needs to be done:

Sending transactions from a 4.11 node to a 4.12 node works, but not the other way round. A new WireTransaction component group needs to be introduced for storing 4.12 attachments so that they can be safely ignored by 4.11 nodes, and the 4.12 node needs to be able to load both 4.11 and 4.12 versions of the same contracts CorDapp so that they can be both attached to the transaction.
Even though attachments are cached when retrieved from the database, the Kotlin metadata version should be stored in the attachments db table, rather than being scanned each time.
Finally, VerificationService was refactored into NodeVerificationSupport and can be passed into SignedTransaction.verifyInternal, instead of needing the much heavier VerifyingServiceHub. This makes it easier for internal tools to verify transactions and spawn the verifier if necessary.

shamsasari commented:
Please merge

adelel1 commented on core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt:
```diff
@@ -191,16 +196,58 @@ data class SignedTransaction(val txBits: SerializedBytes<CoreTransaction>,
+                old = true
+            } else {
+                new = true
+            }
```
Once you find old == true and new == true then you could return straight away without going through all attachments. Maybe a small perf benefit?

shamsasari replied:
Good point. But it looks like this method is going away anyway with the introduction of the new component group.
